packages/mongodb-ts-autocomplete/src/index.ts

```diff
@@ -0,0 +1,206 @@
+import Autocompleter from '@mongodb-js/ts-autocomplete';
+import type {
+  AutocompleterOptions,
+  AutoCompletion,
+} from '@mongodb-js/ts-autocomplete';
+import autocompleteTypes from './fixtures/autocomplete-types';
+
+import type { JSONSchema } from './type-export';
+import { toTypescriptTypeDefinition } from './type-export';
+
+import {
+  inferCollectionNameFromFunctionCall,
+  compileSourceFile,
+} from './cdt-analyser';
+
+import { CachingAutocompletionContext } from './autocompletion-context';
+import type { AutocompletionContext } from './autocompletion-context';
+
+export { JSONSchema } from './type-export';
+export { AutocompletionContext } from './autocompletion-context';
+
+type MongoDBAutocompleterOptions = {
+  context: AutocompletionContext;
+  autocompleterOptions?: AutocompleterOptions;
+};
+
+class DatabaseSchema {
+  private collectionNames: Set<string>;
+  private collectionSchemas: Record<string, JSONSchema>;
+
+  constructor() {
+    // TODO: this is kinda the only real reason for this class: So we can keep
+    // track of the known collectionNames for a database. It could be all the
+    // names from a listCollections() or it could just be all the ones we've
+    // auto-completed for. The schemas can come from the autocompletion context.
+    // This can probably be added to the autocompletion context.
+    this.collectionNames = new Set();
+    this.collectionSchemas = Object.create(null);
+  }
+
+  setCollectionNames(collectionNames: string[]): void {
+    this.collectionNames = new Set(collectionNames);
+  }
+
+  setCollectionSchema(collectionName: string, schema: JSONSchema): void {
+    this.collectionNames.add(collectionName);
+    this.collectionSchemas[collectionName] = schema;
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const collectionProperties = [...this.collectionNames.values()].map(
+      (collectionName) => {
+        const def = this.collectionSchemas[collectionName]
+          ? toTypescriptTypeDefinition(this.collectionSchemas[collectionName])
+          : `{}`;
+        return `      '${collectionName}': ShellAPI.Collection<${def}>;`;
+      }
+    );
+
+    return `{
+  ${collectionProperties.join('\n')}
+}`;
+  }
+}
+
+class ConnectionSchema {
+  private readonly databaseNames: Set<string>;
+  private readonly databaseSchemas: Record<string, DatabaseSchema>;
+
+  constructor() {
+    // TODO: this is kinda the only real reason for this class: So we can keep
```

Comment:
I'll remove the TODO before merging, just keen on hearing people's thoughts first.